### PR TITLE
feat(experimental-utils): move isTypeReadonly from eslint-plugin to experimental-utils

### LINKED ISSUE #3657: Move `isTypeReadonly` from "eslint-plugin" to "experimental-utils".
This util function seems really useful and I'd like to use it in my own rules. Could it be moved to the "experimental-utils" packages so that it is useable by all.

I'll start working on a PR for this now.

re #3305 

Sorry for ignoring the template but none of them seemed appropriate for this issue.

### PR DESCRIPTION
**Pull request by RebeccaStevens**
files moved:
 - isTypeReadonly
 - nullThrows
 - propertyTypes

fix #3657

### CODE REVIEW

File: packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts [hunk 1]
```diff
@@ -3,9 +3,14 @@ import {
...
-import { getFunctionHeadLoc } from './getFunctionHeadLoc';
-import { nullThrows, NullThrowsReasons } from './nullThrows';
+import {
+  isTypeAssertion,
+  isConstructor,
+  isSetter,
+  getFunctionHeadLoc,
+  nullThrows,
+  NullThrowsReasons,
+} from '.';
```
Comment (bradzacher) [inline packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts:13]: we should avoid importing from `.` as it creates a cyclic dependency tree.
eg: `index` imports `explicitReturnTypeUtils` imports `index`.

File: packages/experimental-utils/typings/typescript.d.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import 'typescript';
```
Comment (bradzacher) [inline packages/experimental-utils/typings/typescript.d.ts:1]: do we still need to keep `packages/eslint-plugin/typings/typescript.d.ts` now that we've moved the utils here?

If we still need it - do we need *all* of it, or can we delete parts of it?

**Review by bradzacher — CHANGES_REQUESTED** [review]
Hmm.. This might be a problem for some of our consumers.
Many consumers of `@typescript-eslint/experimental-utils` do so in environments that don't actually use typescript.

for example `eslint-plugin-jest` can be used by non-TS projects, but because the plugin is written in TS they use `experimental-utils`.

Could you please test two things for me:

### (1) do a raw install of the plugin into a folder with no `typescript` dependency.

eg
```ts
cd /tmp
mkdir ts-eslint-test-3658
cd ts-eslint-test-3658
yarn add file:~/path/to/your/checkout/packages/experimental-utils
```

I believe that this should be the correct command to add a local path, and it should show warnings. I think.

The idea is that this should install fine with **no warnings** about missing dependencies.


### (2) make sure it doesn't crash when imported from an environment without TS installed.

```ts
cd /tmp/ts-eslint-test-3658
node --eval="require('@typescript-eslint/experimental-utils')"
```

This should not crash (I might have got the node command wrong... also try `node` and then executing the require in the REPL).

This should execute without crashing.

File: packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts (hunk 1 above)
Comment (RebeccaStevens) [reply packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts]: fixed

File: packages/experimental-utils/typings/typescript.d.ts (hunk 2 above)
Comment (RebeccaStevens) [reply packages/experimental-utils/typings/typescript.d.ts:1]: removed unneeded parts

**Comment by RebeccaStevens** [comment]
I added `typescript` and `tsutils` as optional peer dependencies and conditionally loaded them when needed so that environments with typescript don't have issues.

**Review by bradzacher — COMMENTED** [review]
Thinking through this again - I think that this is probably not the best approach for us to take - if we want to move more things into this package then we're going to have to keep on adding these `ts instanceof Error` checks everywhere - which ultimately makes the codebase really hard to work with.

Perhaps instead we need to create a separate utils package which can be installed by plugins that want to use type-aware utils.

WDYT?

**Comment by RebeccaStevens** [comment]
Yeah, I agree.

Could it be easily made so that all type-aware utils are just put into a subpackage rather than needing a whole new package?

**Comment by bradzacher** [comment]
> Could it be easily made so that all type-aware utils are just put into a subpackage rather than needing a whole new package?

Unfortunately not because node doesn't do dependencies that way - deps are set at the package level only.
The best we could do would be create an import path like `@typescript-eslint/experimental-utils/type-aware` - but the problem there is that currently our build system is dumb and doesn't support multiple entrypoints (which is why all of `experimental-utils` is only available from the top-level export right now).

TS4.5 would make it easy for us to support multiple entrypoints thanks to support for `exports` in package.json (https://github.com/microsoft/TypeScript/issues/33079#issuecomment-913849547) - but that's OFC not due for ~3 months.

A new package would let us not worry about any of this AND also let us ensure we scope our dependencies completely - which does make it seem like the best approach considering we want dependencies on TS and `ts-utils`. There's probably a number of type-aware utils from eslint-plugin we could move across with it to enrich the entire community!

A new package is easy peasy to do - just copy the structure of an existing package. We can go with a name like `type-utils` (I kind of want to go with the name `experimental-type-utils` as a joke because `experimental-utils` was supposed to be renamed back with the 2.0 release - but I guess we should be a little bit serious and "professional" 😉).

**Comment by RebeccaStevens** [comment]
Haha, sounds good to me.

Do you want me to create the package?

**Comment by bradzacher** [comment]
Yes please!
Please update this PR to move all the files to a new package within our `packages` sir and create all the basic (mostly empty) stuff (readme, jest config, tsconfigs, package.json). Most if it can be copied and edited from `experimental-utils`.

Once you've done that I'll hop in and push a commit to add in anything you miss (to save us going back and forth).

Once we merge this we'll take the manual step to create the package in NPM - from there it'll just get pushed automatically in future with our weekly release!

**Comment by bradzacher** [comment]
Let's keep it small for now and we can build on it in future PRs!

**Comment by RebeccaStevens** [comment]
Sounds good.
Thanks 😄

**Comment by bradzacher** [comment]
update: I haven't forgotten about this - on my todo list. Hopefully this weekend.

**Comment by RebeccaStevens** [comment]
@bradzacher Updated again.

**Comment by bradzacher** [comment]
sorry it took me so long to get back to this!
I rebased, fixed the integration tests, and moved more of our type utils across.
I'll merge this as soon as the checks pass!

**Review by bradzacher — DISMISSED** [review]

**Review by bradzacher — DISMISSED** [review]

**Pull request merged**